framework/src/bcs/ArrayLowerDIntegratedBC.C

```diff
@@ -98,7 +98,13 @@ ArrayLowerDIntegratedBC::computeLowerDJacobian(Moose::ConstraintJacobianType typ
                           ? _lowerd_var.number()
                           : _var.number();
 
-  prepareMatrixTagLower(_assembly, ivar, jvar, type);
+  // need to transform the type for assembling Jacobian on boundary to be consistent with
+  // Assembly::addJacobianLowerD() and Assembly::prepareLowerD().
+  Moose::ConstraintJacobianType type_tr =
+      (type == Moose::LowerLower
+           ? type
+           : (type == Moose::LowerPrimary ? Moose::LowerSecondary : Moose::SecondaryLower));
```

Comment:
This seems really wonky. Can't you just modify the called `Moose::ConstraintJacobianTypes` in `ArrayLowerDIntegratedBC::computeJacobian()`?

Reply:
In mortar, which is where a lot of the code you are reusing comes from, the LM variable actually lives on the secondary side of the mortar interface so if reusing that code doing calls to `Moose::LowerLower, Moose::LowerSecondary, Moose::SecondaryLower` would be very natural

Reply:
The problem is that I want to expose users `Moose::LowerLower`, `Moose::LowerPrimary` and `Moose::PrimaryLower`. On the boundary, we only have one element, calling it secondary does not make much sense, do not you agree?

Reply:
Because we cannot modify
```
    if (lower_d_elem)
      _fe_problem.reinitLowerDElem(lower_d_elem, _tid);
```
in `ComputeJacobianThread::onBoundary`, so our option is limited.

Reply:
> On the boundary, we only have one element, calling it secondary does not make much sense, do not you agree?

It depends on the user's background. If they have a constraint or mortar background then this terminology should make good sense to them. In constraint problems it's very common that a secondary (what we used to call slave) node doesn't project onto a primary (what we used to call master) face. If we had chosen different terminlogy such as as `Worker` and `Director` perhaps we would not have any dispute here.

In a vacuum, I agree that if one just encountered a problem in which there is one higher-dimensional element, then presenting only `Primary` (given the language we've chosen) makes sense.

Reply:
> Because we cannot modify
> 
> ```
>     if (lower_d_elem)
>       _fe_problem.reinitLowerDElem(lower_d_elem, _tid);
> ```
> 
> in `ComputeJacobianThread::onBoundary`, so our option is limited.

It's not clear to me why this is relevant